### PR TITLE
Feature/4 edml support for configuration

### LINKED ISSUE #4: edml support for additional configuration
add support for additional configuration (in this case on the file level)

### PR DESCRIPTION
**Pull request by pj-spoelders**
Closes #4

### CODE REVIEW

File: doc/changes/changes_1.1.0.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+Code name: EDML 1.4 / Support for additional configuration
+
+## Summary
+
+EDML 1.4 / Support for additional configuration
+
+## Features
+
+* Add EDML 1.4 Schema (with optional json object for additional configuration)
```
Comment (kaklakariada) [inline doc/changes/changes_1.1.0.md]: Add issue number
```suggestion
* #4 Add EDML 1.4 Schema (with optional json object for additional configuration)
```

File: src/main/java/com/exasol/adapter/document/edml/deserializer/DeserializationHelper.java [hunk 2]
```diff
@@ -56,4 +56,7 @@ static Optional<Boolean> readOptionalBoolean(final JsonObject json, final String
             return Optional.empty();
         }
     }
+    public static String jsonObjectToString(JsonObject jsonObject){
+return jsonObject.toString();
```
Comment (ckunki) [inline src/main/java/com/exasol/adapter/document/edml/deserializer/DeserializationHelper.java]: Strange indent .... 
Maybe formatter is not activated?

File: pom.xml [hunk 3]
```diff
@@ -1,13 +1,12 @@
...
     <artifactId>edml-java</artifactId>
-    <version>1.0.0</version>
+    <version>1.1.0</version>
     <name>EDML Java</name>
     <description>Library for reading and writing EDML definitions.</description>
-    <url>https://github.com/exasol/edml-java</url>
+    <url>https://github.com/exasol/edml-java/</url>
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
```
Comment (kaklakariada) [inline pom.xml]:
```suggestion
```
Remove this and the following properties already defined in the parent pom

File: release_config.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+release-platforms:
+  - GitHub
+  - Maven
+  - Jira
```
Comment (kaklakariada) [inline release_config.yml]: Do we need to announce this for customers in Jira?

File: src/main/java/com/exasol/adapter/document/edml/EdmlKeys.java [hunk 5]
```diff
@@ -16,6 +16,10 @@ public class EdmlKeys {
      * The constant KEY_DESTINATION_TABLE.
      */
     public static final String KEY_DESTINATION_TABLE = "destinationTable";
+    /**
+     * The constant KEY_SOURCE.
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/document/edml/EdmlKeys.java]:
```suggestion
     * The constant KEY_ADDITIONAL_CONFIGURATION.
```

File: src/main/java/com/exasol/adapter/document/edml/deserializer/DeserializationHelper.java [hunk 6]
```diff
@@ -56,4 +56,7 @@ static Optional<Boolean> readOptionalBoolean(final JsonObject json, final String
...
+    }
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/document/edml/deserializer/DeserializationHelper.java]:
```suggestion
    public static String jsonObjectToString(JsonObject jsonObject){
        return jsonObject.toString();
    }
```

File: src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java [hunk 7]
```diff
@@ -31,6 +31,9 @@ public EdmlDefinition deserialize(final String edmlDefinitionAsJson) {
                 .destinationTable(readRequiredString(json, KEY_DESTINATION_TABLE));
         Optional.ofNullable(json.getString(KEY_DESCRIPTION, null)).ifPresent(builder::description);
         readOptionalBoolean(json, KEY_ADD_SOURCE_REFERENCE_COLUMN).ifPresent(builder::addSourceReferenceColumn);
+        //we serialize the json object again, we can then just deserialise it again to a jsonobject later and read it out where we need it
+        JsonObject additionalConfiguration  = json.getJsonObject(KEY_ADDITIONAL_CONFIGURATION); // will return null if not found
+        Optional.ofNullable(additionalConfiguration).ifPresent(addConfig -> builder.additionalConfiguration(jsonObjectToString(addConfig)));
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java]: 👍

**Review by ckunki — DISMISSED** [review]

File: src/main/java/com/exasol/adapter/document/edml/serializer/EdmlSerializer.java [hunk 8]
```diff
@@ -21,7 +22,7 @@
  */
 //[impl->dsn~edml-serialization~1]
 public class EdmlSerializer {
-    private static final String SCHEMA = "https://schemas.exasol.com/edml-1.3.0.json";
+    private static final String SCHEMA = "https://schemas.exasol.com/edml-1.4.0.json";
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/document/edml/serializer/EdmlSerializer.java:27]: Duplicate definition, the version is already defined in `EdmlSchemaValidator.MAPPING_LANGUAGE_SCHEMA`. Maybe use the constant defined there.

File: src/test/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializerTest.java [hunk 9]
```diff
@@ -85,4 +89,15 @@ private EdmlDefinition getEdmlDefinitionForMapping(final MappingDefinition mappi
         return EdmlDefinition.builder().source("test").destinationTable("test").mapping(Fields.builder()//
                 .mapField("myProperty", mapping).build()).build();
     }
+
+    @Test
+    void testHelperJsonObjectToString() {
+        var jsonInput =        "{\"additionalConfiguration\":{\"csv-headers\":true}}";
```
Comment (kaklakariada) [inline src/test/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializerTest.java]:
```suggestion
        var jsonInput = "{\"additionalConfiguration\":{\"csv-headers\":true}}";
```

**Review by kaklakariada — DISMISSED** [review]

File: doc/changes/changes_1.1.0.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
+# EDML Java 1.1.0, released 2022-06-24
```
Comment (kaklakariada) [inline doc/changes/changes_1.1.0.md]: Update release date once you are ready :)

File: src/main/java/com/exasol/adapter/document/edml/deserializer/DeserializationHelper.java (hunk 2 above)
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/document/edml/deserializer/DeserializationHelper.java]: good point, didn't run the formatter on  this file

File: release_config.yml (hunk 4 above)
Comment (pj-spoelders) [reply release_config.yml]: I'll remove it

File: src/main/java/com/exasol/adapter/document/edml/serializer/EdmlSerializer.java (hunk 8 above)
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/document/edml/serializer/EdmlSerializer.java:27]: Values differ / Used for different things. I'll leave this one alone

File: src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java [hunk 11]
```diff
@@ -1,10 +1,10 @@
 package com.exasol.adapter.document.edml.deserializer;
 
 import static com.exasol.adapter.document.edml.EdmlKeys.*;
-import static com.exasol.adapter.document.edml.deserializer.DeserializationHelper.readOptionalBoolean;
-import static com.exasol.adapter.document.edml.deserializer.DeserializationHelper.readRequiredString;
+import static com.exasol.adapter.document.edml.deserializer.DeserializationHelper.*;
 
 import java.io.StringReader;
+import java.io.StringWriter;
```
Comment (ckunki) [inline src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java]: never used. Proposal: organize imports.

File: src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java [hunk 12]
```diff
@@ -31,6 +31,9 @@ public EdmlDefinition deserialize(final String edmlDefinitionAsJson) {
                 .destinationTable(readRequiredString(json, KEY_DESTINATION_TABLE));
         Optional.ofNullable(json.getString(KEY_DESCRIPTION, null)).ifPresent(builder::description);
         readOptionalBoolean(json, KEY_ADD_SOURCE_REFERENCE_COLUMN).ifPresent(builder::addSourceReferenceColumn);
+        //we serialize the json object again, we can then just deserialise it again to a json object later and read it out where we need it
+        JsonObject additionalConfiguration = json.getJsonObject(KEY_ADDITIONAL_CONFIGURATION); // will return null if not found
+        Optional.ofNullable(additionalConfiguration).ifPresent(addConfig -> builder.additionalConfiguration(jsonObjectToString(addConfig)));
```
Comment (ckunki) [inline src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java]:
```suggestion
        if(additionalConfiguration != null) {
           builder.additionalConfiguration(jsonObjectToString(additionalConfiguration));
        }
```

File: src/main/java/com/exasol/adapter/document/edml/deserializer/MappingDeserializer.java [hunk 13]
```diff
@@ -13,23 +13,23 @@
 
 class MappingDeserializer {
     static void deserializeToColumnMapping(final JsonObject json,
-            final AbstractToColumnMapping.AbstractToColumnMappingBuilder<?, ?> builder) {
+                                           final AbstractToColumnMapping.AbstractToColumnMappingBuilder<?, ?> builder) {
```
Comment (ckunki) [inline src/main/java/com/exasol/adapter/document/edml/deserializer/MappingDeserializer.java]: Could it be that your formatter settings differ from default?

File: src/test/java/com/exasol/adapter/document/edml/MappingTestFiles.java [hunk 14]
```diff
@@ -54,7 +54,7 @@ public static String getMappingAsString(final String fileName) {
      * @throws IOException on read or write error.
      */
     public static File generateInvalidFile(final String baseMappingName,
-            final Function<JSONObject, JSONObject> invalidator, final Path tempDir) throws IOException {
+                                           final Function<JSONObject, JSONObject> invalidator, final Path tempDir) throws IOException {
```
Comment (ckunki) [inline src/test/java/com/exasol/adapter/document/edml/MappingTestFiles.java]: formatter?

**Review by ckunki — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java (hunk 11 above)
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java]: done

File: src/main/java/com/exasol/adapter/document/edml/deserializer/MappingDeserializer.java (hunk 13 above)
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/document/edml/deserializer/MappingDeserializer.java]: It was. Good you noticed. Fixed this now.

File: src/test/java/com/exasol/adapter/document/edml/MappingTestFiles.java (hunk 14 above)
Comment (pj-spoelders) [reply src/test/java/com/exasol/adapter/document/edml/MappingTestFiles.java]: Yep ..

File: src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java (hunk 12 above)
Comment (pj-spoelders) [reply src/main/java/com/exasol/adapter/document/edml/deserializer/EdmlDeserializer.java]: Fair enough ..

**Review by ckunki — APPROVED** [review]

**Pull request merged**